lens/registry.go

```diff
@@ -48,59 +48,149 @@ type lensRegistry struct {
 
 	lensPoolsBySchemaVersionID     map[string]*lensPool
 	reversedPoolsBySchemaVersionID map[string]*lensPool
+	poolLock                       sync.RWMutex
 
 	// lens configurations by source schema version ID
-	configs map[string]client.LensConfig
+	configs    map[string]client.LensConfig
+	configLock sync.RWMutex
+
+	// Writable transaction contexts by transaction ID.
+	//
+	// Read-only transaction contexts are not tracked.
+	txnCtxs map[uint64]*txnContext
+	txnLock sync.RWMutex
```

Comment:
I think you'd actually only need to keep it in `explicitTxnLensRegistry`

Reply:
We couldn't delete any code if we moved this to `explicitTxnLensRegistry`, it would just move. What do you think would get simpler?

If either one of `implicitTxnLensRegistry` or `explicitTxnLensRegistry` needs the txnCtx then most of the functions in registry require it in order to keep the func signatures the same (or we'd have to duplicate a bunch of code)

Reply:
`getCtx` gets much simpler and no need to worry about transaction IDs: https://github.com/fredcarle/defradb/commit/b059e912fa9acf38a1108a93ba632372a0b4a136

Reply:
I think you have made the mistake I made at first, doing it like that means that the same txn cannot scope across many different LensRegistry instances, and I belive the tests do catch this and the changes you linked should fail the build.

For example:    
```
txn := db.NewTxn()
reg1 := db.WithTxn(txn).LensRegistry()
reg2 := db.WithTxn(txn).LensRegistry()

// reg1 and reg do not share scope with your proposed method. But they absolutely have to, as they are scoped to the same txn
```

Reply:
In your example though it's the same LensRegistry instance so it should work fine. The operations to the registry get the txnContext as a parameter anyways.

Reply:
They are different instances

Reply:
Also, the current (registry) solution should still work when we allow external management of transaction, whereas yours would require us to revert your suggestion and re-introduce the one currently within the PR.

Datastore transactions are not restricted to a single in-memory instance.  Long term we will need to change the ID logic to properly reflect this (e.g. for badger we should perhaps use the Badger txn id/ts), but the lens registry stuff should be able to stay the same.

Reply:
I was thinking of the lensRegistry struct. This one is the same but yes the explicitTxnLensRegistry is different.

`re2.Config` will be different because the txnContext is different.

Reply:
Alright, thanks for your engagement here. I think I can see the bigger picture more clearly. 

Side thought: Having external management of transactions will make DefraDB quite in a league of it's own where software interfacing with DefraDB won't need a driver to be able to do that kind of interaction. Which is pretty cool.